pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,433 @@
+			c.createPodInformer()
+			go c.podInformer.Run(c.podWorkerStopCh)
+			if !cache.WaitForNamedCacheSync(podRouteControllerName, c.podWorkerStopCh, c.podInformer.HasSynced) {
+				close(c.podWorkerStopCh)
+				c.podInformer = nil
+				c.podLister = nil
+				return errors.New("failed to sync Pod cache")
+			}
+
+			for i := 0; i < podWorkerNum; i++ {
+				go wait.Until(c.podWorker, time.Second, c.podWorkerStopCh)
+			}
+			c.podWorkersStarted = true
+			return nil
+		}
+		// Do nothing when the Pod flow controller is already started since
+		// Pod flow controller will be responsible for handling Pod events to install flows.
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) podWorker() {
+	for c.processPodNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processPodNextWorkItem() bool {
+	obj, quit := c.podQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.podQueue.Done(obj)
+
+	if k, ok := obj.(string); !ok {
+		c.podQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncPod(k); err == nil {
+		c.podQueue.Forget(k)
+	} else {
+		c.podQueue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) syncPod(podIP string) error {
+	c.podWorkersStartedMutex.RLock()
+	defer c.podWorkersStartedMutex.RUnlock()
+	if !c.podWorkersStarted {
+		return nil
+	}
+
+	podSet := c.getPodsByIP(podIP)
+	if podSet.Len() == 0 {
+		klog.V(2).InfoS("Deleting Multi-cluster flows for PodIP", "podIP", podIP)
+		if err := c.ofClient.UninstallMulticlusterPodFlows(podIP); err != nil {
+			klog.ErrorS(err, "Failed to uninstall Multi-cluster flows for Pod", "podIP", podIP)
+			return err
+		}
+		return nil
+	}
+
+	latestIPs, exists := c.getPodNodeIPs(podIP)
+	if !exists {
+		return nil
+	}
+
+	klog.V(2).InfoS("Adding Multi-cluster flows for PodIP", "podIP", latestIPs.podIP, "nodeIP", latestIPs.podNodeIP)
+	if err := c.ofClient.InstallMulticlusterPodFlows(net.ParseIP(latestIPs.podIP), net.ParseIP(latestIPs.podNodeIP)); err != nil {
+		klog.ErrorS(err, "Failed to install Multi-cluster flows for Pod", "podIP", latestIPs.podIP, "nodeIP", latestIPs.podNodeIP)
+		return err
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) getPodsByIP(ip string) sets.String {
+	c.ipToPodsMutex.RLock()
+	defer c.ipToPodsMutex.RUnlock()
+	podSet, ok := c.ipToPods[ip]
+	if ok {
+		return podSet
+	}
+	return nil
+}
+
+func (c *MCPodRouteController) addPodCache(ip string, podName types.NamespacedName) {
+	c.ipToPodsMutex.Lock()
+	defer c.ipToPodsMutex.Unlock()
+	if pods, ok := c.ipToPods[ip]; !ok {
+		c.ipToPods[ip] = sets.NewString(podName.String())
+	} else {
+		pods.Insert(podName.String())
+	}
+}
+
+func (c *MCPodRouteController) deletePodCache(ip string, podName types.NamespacedName) {
+	c.ipToPodsMutex.Lock()
+	defer c.ipToPodsMutex.Unlock()
+	podSet := c.ipToPods[ip]
+	podSet.Delete(podName.String())
```

Comment:
When `podSet` is empty, we should delete the podIP from `c.ipToPods`

Reply:
Yes, the deletion is happened after `c.deletePodNodeIP` is called.
```
			if podSet.Len() == 1 {
				c.deletePodNodeIP(podIP)
			}
			c.deletePodCache(podIP, podName)
```

Reply:
？ When you delete `podIP `from `c.ipToPods`? `deletePodNodeIP `just deletes it from the `podCache`

Reply:
Miss-understood it, added following check to clean up:
```
	if podSet.Len() == 0 {
		delete(c.ipToPods, ip)
	}
```